GitUI/Shells/ShellProvider.cs

```diff
@@ -0,0 +1,43 @@
+﻿using ConEmu.WinForms;
+
+namespace GitUI.Shells
+{
+    public class ShellProvider
+    {
+        private static PwshShell PwshShellInstance;
+        private static PowerShellShell PowerShellShellInstance;
+        private static CmdShell CmdShellInstance;
+        private static BashShell BashShellInstance;
+
+        public IShellDescriptor GetShell(string shellType)
+        {
+            switch (shellType.ToLowerInvariant())
```

Comment:
I'm not 100% sold by the introduction of `ShellType' that:
* seems redundant with names
* makes (a little) more difficult to add a new shell because you should update 2 places (ShellType and here) and there is no help for the developer if there is a mismatch

The only advantage is to be able to so `ShellType.Bash` but for this special case, a `const` could be used.

I will push a commit with a draft in another branch https://github.com/pmiossec/gitextensions/tree/refactor_shells
Please, have a look....

Reply:
The idea with the enum was to convert the setting to the enum. But your version offers probably equally good alternative. 
I left a comment there.